graph/topo/johnson_cycles.go

```diff
@@ -70,6 +70,114 @@ func DirectedCyclesIn(g graph.Directed) [][]graph.Node {
 	return j.result
 }
 
+// DirectedCyclesContaining returns the set of elementary cycles in the graph g containing the node with ID vid.
+// This is a slight deviation from Johnson's.
+func DirectedCyclesContaining(g graph.Directed, vid int64) [][]graph.Node {
+	jg := johnsonGraphFrom(g)
+	j := johnson{
+		adjacent: jg,
+		b:        make([]set.Ints, len(jg.orig)),
+		blocked:  make([]bool, len(jg.orig)),
+	}
+
+	j.s = j.adjacent.indexOf(vid)
+	// We use the entire SCC adjacency to find all cycles, not just those with least vertex >= s.
+	sccs := TarjanSCC(j.adjacent)
+	// A_k = adjacency structure of (maximal) strong component K containing s
+	j.adjacent = j.adjacent.sccSubGraph(sccs, 2) // Only allow SCCs with >= 2 vertices.
+	if j.adjacent.order() == 0 {
+		return j.result
+	}
+
+	for i, v := range j.adjacent.orig {
+		if !j.adjacent.nodes.Has(v.ID()) {
+			continue
+		}
+		if len(j.adjacent.succ[v.ID()]) > 0 {
+			j.blocked[i] = false
+			j.b[i] = make(set.Ints)
+		}
+	}
```

Comment:
Can you factor this out to a helper so it's not quadruplicated?

Reply:
If you don't mind, a question I have is if and when this conditional is true https://github.com/gonum/gonum/blob/eae99adef421605fe44c4cab6e45a17e7e846560/graph/topo/johnson_cycles.go#L53. I know that it is in the original paper as `s := least vertex in V_k;` but I don't see how `j.adjacent.leastVertexIndex()` could return a number less than `j.s`.

Reply:
Yes, it can be true, both as an equality (which doesn't matter) and as `s > j.s`. This happens when the least index is not in V_k.